File: Contests/BiWeekly/27/2_check_if_a_string_contains_all_binary_codes_of_size_k.py
from collections import defaultdict


class Solution:
    def hasAllCodes(self, s: str, k: int) -> bool:
        d = defaultdict(int)
        for i in range(len(s)-k+1):
            d[s[i:i+k]] += 1
        if len(d) != 2**k:
            return False
        return True

File: Contests/BiWeekly/27/2_check_if_a_string_contains_all_binary_codes_of_size_k.cpp
class Solution {
public:
    bool hasAllCodes(string s, int k) {
        int n = s.size();
        unordered_set<string> visited;
        if(k > n) return false;
        int i = n - k;
        string temp = s.substr(i);
        visited.insert(temp);
        i--;
        //Going in reverse because CPP 
        //strings do not have O(1) pop from start
        for(; i >= 0; i--){
            temp = s[i] + temp;
            temp.pop_back();
            visited.insert(temp);
        }
        return visited.size() == 1<<k;
    }
};